### ISSUE TITLE
Nested defaultdict loses its default factory

### ISSUE THREAD

**Issue by Thoufak** [pydantic/pydantic#4687]
### Initial Checks

- [X] I have searched GitHub for a duplicate issue and I'm sure this is something new
- [X] I have searched Google & StackOverflow for a solution and couldn't find anything
- [X] I have read and followed [the docs](https://pydantic-docs.helpmanual.io) and still think this is a bug
- [X] I am confident that the issue is with pydantic (not my code, or another library in the ecosystem like [FastAPI](https://fastapi.tiangolo.com) or [mypy](https://mypy.readthedocs.io/en/stable))

### Description

Pydantic fails to properly construct the defaultdict when it is nested into another defaultdict.
After the model construction, the outer dict will have the correct default_factory, but the inner one will have its factory set to None.

### Example Code

```Python
from collections import defaultdict
from typing import DefaultDict
from pydantic import BaseModel


class TestModel(BaseModel):
    d: DefaultDict[int, DefaultDict[int, int]]


model = TestModel(d=defaultdict(lambda: defaultdict(int)))
parsed = TestModel.parse_raw(model.json())
print(parsed.d[1])     # defaultdict(None, {})
print(parsed.d[1][2])  # KeyError: 2
```

### Python, Pydantic & OS Version

```Text
pydantic version: 1.9.1
            pydantic compiled: True
                 install path: /Users/abc/PycharmProjects/Pybugtic/resources/venv/lib/python3.9/site-packages/pydantic
               python version: 3.9.5 (default, Jul  9 2022, 23:20:38)  [Clang 13.1.6 (clang-1316.0.21.2.5)]
                     platform: macOS-12.3-arm64-arm-64bit
     optional deps. installed: ['typing-extensions']
```

### Affected Components

- [ ] [Compatibility between releases](https://pydantic-docs.helpmanual.io/changelog/)
- [X] [Data validation/parsing](https://pydantic-docs.helpmanual.io/usage/models/#basic-model-usage)
- [ ] [Data serialization](https://pydantic-docs.helpmanual.io/usage/exporting_models/) - `.dict()` and `.json()`
- [ ] [JSON Schema](https://pydantic-docs.helpmanual.io/usage/schema/)
- [ ] [Dataclasses](https://pydantic-docs.helpmanual.io/usage/dataclasses/)
- [ ] [Model Config](https://pydantic-docs.helpmanual.io/usage/model_config/)
- [ ] [Field Types](https://pydantic-docs.helpmanual.io/usage/types/) - adding or changing a particular data type
- [ ] [Function validation decorator](https://pydantic-docs.helpmanual.io/usage/validation_decorator/)
- [ ] [Generic Models](https://pydantic-docs.helpmanual.io/usage/models/#generic-models)
- [ ] [Other Model behaviour](https://pydantic-docs.helpmanual.io/usage/models/) - `construct()`, pickling, private attributes, ORM mode
- [ ] [Plugins](https://pydantic-docs.helpmanual.io/) and integration with other tools - mypy, FastAPI, python-devtools, Hypothesis, VS Code, PyCharm, etc.

**Comment by Thoufak** [comment]
A possible workaround is to add a validator like this:
```python
class TestModel(BaseModel):
    d: DefaultDict[int, DefaultDict[int, int]]

    @validator("d")
    def dd(cls, value):
        value.default_factory = lambda: defaultdict(int)
        return value
```

**Comment by samuelcolvin** [comment]
I believe this is fixed by #4516 and will be fixed in v2.

**Comment by adriangb** [comment]
Resolved in #5744
